### ISSUE TITLE
False positives for content-type meta element presence check

### ISSUE THREAD

**Issue by bbeckford** [MyIntervals/emogrifier#957]
I've recently updated Emogrifier and when moving from the old style syntax to the new, I receive this error:
`PHP Fatal error:  Uncaught TypeError: Return value of Pelago\Emogrifier\CssInliner::getHeadElement() must be an instance of DOMElement, null returned in /var/www/vhosts/test.system/httpdocs/vendor/pelago/emogrifier/src/CssInliner.php:1293`

Is there a reason there must always be a `<head>` element?
https://github.com/MyIntervals/emogrifier/blob/b43b650880d189b0ada61d95d0729c7424b1752d/src/CssInliner.php#L1291

I'm just inlining CSS in a simple email template that doesn't use a `<head>` element for anything.
I've added a blank one for now but strikes me as a little arbitrary.

Thanks!

**Comment by JakeQZ** [comment]
Hi @bbeckford,

Thanks for the report.

This is likely to be a bug and may be similar to #923.  If this was working with the previous version then it may have been introduced by #853 (which was first included in 4.0.0).  Emogrifier is supposed to add a `<head>` element if there isn't one (and also a `Content-Type`), so that by the time `getHeadElement()` is called, there is one.

Could you post your HTML?

**Comment by JakeQZ** [comment]
#642 (first included in 2.1.1) may also have introduced the problem.  Prior to this, although the `<head>` element was supposed to be added if missing, Emogrifier did not depend on it because the `<style>` element was previously added to `<body>`.

**Comment by bbeckford** [comment]
Thanks for the response.

To be fair we were using version 0.1.1 before!

Sure here you go:

    <!DOCTYPE html>
    <html>
        <body class="cpTemEmail cp_standard_email">

    <div class="cpHeader">
        <div class="cpHeaderRow first">
            <div class="ccd_logo" style="background-image: none;"><a href="http://www.redacted.link" target="_blank"><img class="email_logo" src="cid:ccd_logo" width="70" height="23"></a></div>
            <p class="editedDate"><span>27th Jan 2021</span> at <span>19:48</span></p>
            <h1>test Form Received #31696</h1>
            <div style="clear:both;"></div>
        </div>
    </div>
    <div class="cpWrapper">
        <div class="cpMessageBody">
                <div class="avoidPageBreak">
            <div class="pageBreakSpacer"></div>
            <div class="sectionWrapper top_level" id="divider-1">
                <p class="dividerTitle">test</p>
                <table>
                    <tbody>
                        <tr>
                            <td><p>test</p></td>
                            <td><p>sdfsdf</p></td>
                        </tr>
                        <tr>
                            <td><p>9 chars test</p></td>
                            <td><p>sdfsdfsdf</p></td>
                        </tr>
                        <tr>
                            <td><p>10 digits test</p></td>
                            <td><p>3453453453</p></td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </div>

        </div>
    </div> </body>
    </html>

**Comment by JakeQZ** [comment]
Hmm, I can't reproduce a problem with this HTML against the latest `main` (and don't spot anything that we know from #923 might cause a problem, such as an invalidly placed `Content-Type`, or that previously caused a problem, such as `<header>` elements).

What version of Emogrifier are you upgrading to?

You mentioned you are using a template.  Could you check that this specific HTML actually fails, or perhaps it is some specific HTML inserted through the template that causes the problem?

I added the following test, which passed, but should fail with HTML which exhibits the problem:

```php
    /**
     * @test
     */
    public function test957()
    {
        $subject = CssInliner::fromHtml(<<<EOHTML
<!DOCTYPE html>
<html>
    <body class="cpTemEmail cp_standard_email">

<div class="cpHeader">
    <div class="cpHeaderRow first">
        <div class="ccd_logo" style="background-image: none;"><a href="http://www.redacted.link" target="_blank"><img class="email_logo" src="cid:ccd_logo" width="70" height="23"></a></div>
        <p class="editedDate"><span>27th Jan 2021</span> at <span>19:48</span></p>
        <h1>test Form Received #31696</h1>
        <div style="clear:both;"></div>
    </div>
</div>
<div class="cpWrapper">
    <div class="cpMessageBody">
            <div class="avoidPageBreak">
        <div class="pageBreakSpacer"></div>
        <div class="sectionWrapper top_level" id="divider-1">
            <p class="dividerTitle">test</p>
            <table>
                <tbody>
                    <tr>
                        <td><p>test</p></td>
                        <td><p>sdfsdf</p></td>
                    </tr>
                    <tr>
                        <td><p>9 chars test</p></td>
                        <td><p>sdfsdfsdf</p></td>
                    </tr>
                    <tr>
                        <td><p>10 digits test</p></td>
                        <td><p>3453453453</p></td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>

    </div>
</div> </body>
</html>
EOHTML
        );
        //$subject->setDebug(true); // makes no difference

        // `:hover` selector can't be inlined so ensures `<style>` element will be added to `<head>`
        $subject->inlineCss('div { color: green; } div:hover { color: blue; }');

        $result = $subject->render();
        self::assertStringContainsString('color: green', $result);
        self::assertStringContainsString('color: blue', $result);
        \error_log("Resulting HTML:\n" . $result);
    }
```

The result had the `<head>` element added:

```html
<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">div:hover{color: blue;}</style>
</head>
    <body class="cpTemEmail cp_standard_email">

<div class="cpHeader" style="color: green;">
    <div class="cpHeaderRow first" style="color: green;">
        <div class="ccd_logo" style="color: green; background-image: none;"><a href="http://www.redacted.link" target="_blank"><img class="email_logo" src="cid:ccd_logo" width="70" height="23"></a></div>
        <p class="editedDate"><span>27th Jan 2021</span> at <span>19:48</span></p>
        <h1>test Form Received #31696</h1>
        <div style="color: green; clear: both;"></div>
    </div>
</div>
<div class="cpWrapper" style="color: green;">
    <div class="cpMessageBody" style="color: green;">
            <div class="avoidPageBreak" style="color: green;">
        <div class="pageBreakSpacer" style="color: green;"></div>
        <div class="sectionWrapper top_level" id="divider-1" style="color: green;">
            <p class="dividerTitle">test</p>
            <table>
                <tbody>
                    <tr>
                        <td><p>test</p></td>
                        <td><p>sdfsdf</p></td>
                    </tr>
                    <tr>
                        <td><p>9 chars test</p></td>
                        <td><p>sdfsdfsdf</p></td>
                    </tr>
                    <tr>
                        <td><p>10 digits test</p></td>
                        <td><p>3453453453</p></td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>

    </div>
</div> </body>
</html>
```

**Comment by JakeQZ** [comment]
Hi @bbeckford,

If you have time to follow up on this re my previous comment&hellip;

> What version of Emogrifier are you upgrading to?
>
> Could you check that this specific HTML actually fails, or perhaps it is some specific HTML inserted through the template that causes the problem?

&hellip;could you also post the output from `composer show -p`, as the system environment (e.g. version of `libxml`) may also have a bearing.

TIA :)

**Comment by bbeckford** [comment]
Hi @JakeQZ - sorry for the delay.

Strange, I can't seem to replicate this now! The dev environment I was testing on has moved on now so there's no way for me to know for sure but perhaps I was using an older version at the time,

Now definitely using emogrifier v5.0.0 and the snippet above does work.

I've removed this line from the code I was working on and I will let you know if this ever reoccurs on any of my instances:
![image](https://user-images.githubusercontent.com/2021072/108744480-e0c84280-7531-11eb-851d-c3576df6e838.png)

Thanks for your help!

**Comment by bbeckford** [comment]
@JakeQZ I've hit this error again and can replicate it with this HTML - https://gist.github.com/bbeckford/eea6c7a1a8a4957166050dda413a9af1

Code to reproduce:

    $template_body_processed = CssInliner::fromHtml(file_get_contents("email.html"))->inlineCss($css)->render();

This produces the following error:

    Fatal error: Uncaught TypeError: Return value of Pelago\Emogrifier\CssInliner::getHeadElement() must be an instance of DOMElement, null returned in /var/www/vhosts/redacted/httpdocs/vendor/pelago/emogrifier/src/CssInliner.php on line 1293
    TypeError: Return value of Pelago\Emogrifier\CssInliner::getHeadElement() must be an instance of DOMElement, null returned in /var/www/vhosts/redacted/httpdocs/vendor/pelago/emogrifier/src/CssInliner.php on line 1293

Version info from composer.lock: (running PHP v7.3.27)



            "name": "pelago/emogrifier",
            "version": "v5.0.0",
            "source": {
                "type": "git",
                "url": "https://github.com/MyIntervals/emogrifier.git",
                "reference": "b43b650880d189b0ada61d95d0729c7424b1752d"
            },

**Comment by oliverklee** [comment]
@bbeckford Thanks! Could you also please post the HTML and CSS you're using?

**Comment by bbeckford** [comment]
Hi @oliverklee included html is here - https://gist.github.com/bbeckford/eea6c7a1a8a4957166050dda413a9af1

I haven't included the CSS at this stage as I'd have to comb through it to make sure it was ok for public consumption, let me see if I can reproduce with something simpler...

**Comment by bbeckford** [comment]
@oliverklee I found a small old CSS file that you can pass in that results in the same bug - https://gist.github.com/bbeckford/f44a201af3ac62471afdb7f9574c83a5

**Comment by oliverklee** [comment]
@bbeckford Thanks! We'll have a look at the problem.

**Comment by JakeQZ** [comment]
Many thanks @bbeckford for following up.  I can reproduce this with the attached HTML.

The HTML contains a literal string `Content-Type` for display.  This causes Emogrifier to believe there is a `Content-Type` `<meta>` element and thus a `<head>` element (implicit or otherwise).  The code needs to be more judicious when determining if such an element exists and is valid.  The offending line is far too simplistic:

```php
        $hasContentTypeMetaTag = \stripos($html, 'Content-Type') !== false;
```

So it's not quite the same bug as #923, though is related.

**Comment by JakeQZ** [comment]
```php
        $hasContentTypeMetaTag
            = \preg_match('%<meta(?=\\s)[^>]*\\shttp-equiv=(["\']?+)Content-Type\\g{-1}[\\s/>]%i', $html) === 1;
```

...should do the job.  Could perhaps be optimized to use possessive quantifiers throughout, though would be small fry for something run only once (but might result in catastrophic failure for certain HTML input if not done).

Would need various tests, e.g. single, double and no quotes for attribute value, as well as case incarnations (I believe all keywords are case insensitive) for positive, and imaginative creations for negative.

The validity of the placement of the `<meta>` element would be covered by #923 as a separate issue.  OTTOMY I think it would be best to fix this (#957) one first.

**Comment by bbeckford** [comment]
FYI 'Content-Type' in this instance is from the output of xdebug, so not entirely uncommon

**Comment by bbeckford** [comment]
Hey thanks for the fix @JakeQZ :)
